source/fuzz/transformation_add_copy_memory.cpp

```diff
@@ -0,0 +1,197 @@
+  message_.set_source_id(source_id);
+  message_.set_storage_class(storage_class);
+  message_.set_initializer_id(initializer_id);
+}
+
+bool TransformationAddCopyMemory::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // Check that target id is fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // Check that instruction descriptor is valid.
+  auto* inst = FindInstruction(message_.instruction_descriptor(), ir_context);
+  if (!inst) {
+    return false;
+  }
+
+  // Check that we can insert OpCopyMemory before |instruction_descriptor|.
+  auto iter = fuzzerutil::GetIteratorForInstruction(
+      ir_context->get_instr_block(inst), inst);
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(SpvOpCopyMemory, iter)) {
+    return false;
+  }
+
+  // Check that source instruction exists.
+  auto* source_inst =
+      ir_context->get_def_use_mgr()->GetDef(message_.source_id());
+  if (!source_inst) {
+    return false;
+  }
+
+  // Check that result type of source instruction exists and can be used with
+  // OpCopyMemory.
+  const auto* source_type =
+      ir_context->get_type_mgr()->GetType(source_inst->type_id());
+  if (!source_type || !source_type->AsPointer() ||
+      !CanUsePointeeWithCopyMemory(*source_type->AsPointer()->pointee_type())) {
+    return false;
+  }
+
+  // |storage_class| is either Function or Private.
+  if (message_.storage_class() != SpvStorageClassFunction &&
+      message_.storage_class() != SpvStorageClassPrivate) {
+    return false;
+  }
+
+  auto pointee_type_id = fuzzerutil::GetPointeeTypeIdFromPointerType(
+      ir_context, source_inst->type_id());
+
+  // OpTypePointer with |message_.storage_class| exists.
+  if (!fuzzerutil::MaybeGetPointerType(
+          ir_context, pointee_type_id,
+          static_cast<SpvStorageClass>(message_.storage_class()))) {
+    return false;
+  }
+
+  // Check that |initializer_id| exists and has valid type.
+  const auto* initializer_inst =
+      ir_context->get_def_use_mgr()->GetDef(message_.initializer_id());
+  if (!initializer_inst || initializer_inst->type_id() != pointee_type_id) {
+    return false;
+  }
+
+  // Check that this transformation respects domination rules.
+  const auto* source_block = ir_context->get_instr_block(source_inst);
+  const auto* target_block = ir_context->get_instr_block(inst);
+  assert(target_block && "OpCopyMemory can't be a global instruction");
+
+  // Domination rules are satisfied if we are copying memory from a global
+  // instruction.
+  if (!source_block) {
+    return true;
+  }
+
+  // Check that both source and target instructions are in the same function.
+  if (source_block->GetParent() != target_block->GetParent()) {
+    return false;
+  }
+
+  // Check domination rules.
+  return source_inst != inst &&
+         ir_context->GetDominatorAnalysis(source_block->GetParent())
+             ->Dominates(source_inst, inst);
```

Comment:
I think you could replace all of this with a call to fuzzerutil::IdIsAvailableBeforeInstruction, which also captures the case where you are copying a function parameter (which I don't think you capture here).

Reply:
Fixed in d32f2fd.